File: challenges/2021/07.js
// Part 1
const solvePuzzle1 = (input) => getMinFuel(input, Math.abs);

// Part 2
const calculateDistance2 = (diff, d = Math.abs(diff)) => (d * (d + 1)) / 2;
const solvePuzzle2 = (input) => getMinFuel(input, calculateDistance2);

const getMinFuel = (input, calculateDistance) => {
  const nums = parseInput(input).sort((a, b) => a - b);
  let minSum = Infinity;

  for (let i = nums[nums.length - 1]; i >= nums[0]; i--) {
    let currSum = 0;

    for (let j = 0; j < nums.length; j++) {
      currSum += calculateDistance(nums[j] - i);
      if (currSum > minSum) break;
    }

    minSum = Math.min(minSum, currSum);
  }

  return minSum;
};

const parseInput = (input) => input.split(",").map((n) => Number(n));

require(__dirname + "/../../utils/test.js").test(__filename, __dirname, solvePuzzle1);
require(__dirname + "/../../utils/test.js").test(__filename, __dirname, solvePuzzle2);

File: challenges/2021/07.exs
Code.compile_file(Path.join([__ENV__.file], ["../../../utils/utils.exs"]))

defmodule Challenge do
  @moduledoc false

  def first_result([input]), do: get_min_fuel(input, & &1)

  def second_result([input]), do: get_min_fuel(input, &(&1 * (&1 + 1) / 2))

  defp get_min_fuel(input, calculate_distance) do
    nums =
      input
      |> String.split(",")
      |> Enum.map(&String.to_integer/1)
      |> Enum.sort()

    Enum.reduce(Enum.at(nums, -1)..Enum.at(nums, 0), :i, fn n, acc ->
      nums
      |> Enum.reduce(0, &(&2 + calculate_distance.(abs(n - &1))))
      |> min(acc)
    end)
    |> round()
  end
end

Utils.run(&Challenge.first_result/1, __ENV__.file, "Q1")
Utils.run(&Challenge.second_result/1, __ENV__.file, "Q2")
